crates/matrix-sdk-indexeddb/src/event_cache_store/mod.rs

```diff
@@ -0,0 +1,421 @@
+        // let migration_strategy = self.migration_conflict_strategy.clone();
+        let name = self.name.unwrap_or_else(|| "event_cache".to_owned());
+
+        let serializer = IndexeddbSerializer::new(self.store_cipher);
+        debug!("IndexedDbEventCacheStore: opening main store {name}");
+        let inner = open_and_upgrade_db(&name, &serializer).await?;
+
+        let store = IndexeddbEventCacheStore { name, inner, serializer };
+        Ok(store)
+    }
+}
+
+/// Sometimes Migrations can't proceed without having to drop existing
+/// data. This allows you to configure, how these cases should be handled.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum MigrationConflictStrategy {
+    /// Just drop the data, we don't care that we have to sync again
+    Drop,
+    /// Raise a [`IndexeddbStateStoreError::MigrationConflict`] error with the
+    /// path to the DB in question. The caller then has to take care about
+    /// what they want to do and try again after.
+    Raise,
+    /// Default.
+    BackupAndDrop,
+}
+
+pub struct IndexeddbEventCacheStore {
+    name: String,
+    pub(crate) inner: IdbDatabase,
+    pub(crate) serializer: IndexeddbSerializer,
+}
+
+#[cfg(not(tarpaulin_include))]
+impl std::fmt::Debug for IndexeddbEventCacheStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("IndexeddbEventCacheStore").finish()
+    }
+}
+
+impl IndexeddbEventCacheStore {
+    pub fn builder() -> IndexeddbEventCacheStoreBuilder {
+        IndexeddbEventCacheStoreBuilder::new()
+    }
+}
+
+type Result<A, E = IndexeddbEventCacheStoreError> = std::result::Result<A, E>;
+
+#[cfg(target_arch = "wasm32")]
+macro_rules! impl_event_cache_store {
+    ({ $($body:tt)* }) => {
+        #[async_trait(?Send)]
+        impl EventCacheStore for IndexeddbEventCacheStore {
+            type Error = IndexeddbEventCacheStoreError;
+
+            $($body)*
+        }
+    };
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+macro_rules! impl_state_store {
+    ({ $($body:tt)* }) => {
+        impl IndexeddbEventCacheStore {
+            $($body)*
+        }
+    };
+}
+
+impl_event_cache_store!({
+    async fn handle_linked_chunk_updates(
+        &self,
+        room_id: &RoomId,
+        updates: Vec<Update<Event, Gap>>,
+    ) -> Result<()> {
+        // TODO not sure if this should be a String or JsValue (which I assume is a ByteArray)
+        let hashed_room_id = self.serializer.encode_key_as_string(keys::LINKED_CHUNKS, room_id);
+        let room_id = room_id.to_owned();
+        // let this = self.clone();
+        let tx = self
+            .inner
+            .transaction_on_one_with_mode(keys::LINKED_CHUNKS, IdbTransactionMode::Readwrite)?;
+
+        let object_store = tx.object_store(keys::LINKED_CHUNKS)?;
+
+        for update in updates {
+            match update {
+                Update::NewItemsChunk { previous, new, next } => {
```

Comment:
Once everything else is reviewed/ready it's just a matter of filling in this blanks to get a working implementation I think